builder: implement PullParent option with buildkit

tiborvass opened the pull request:
Signed-off-by: Tibor Vass <tibor@docker.com>

This allows to run `docker build --pull ...` when buildkit is enabled.

Commit:
builder: implement PullParent option with buildkit

Signed-off-by: Tibor Vass <tibor@docker.com>

vdemeester reviewed (approved):
LGTM 🐯

thaJeztah reviewed (approved):
SGTM

ping @tonistiigi PTAL

tiborvass commented:
@vdemeester @thaJeztah I updated this. I fixed an issue whereby "local" would error out if local was not found, when in fact it just means "prefer" local, but still fallback to remote if cannot find. Ideally I'd like "pull" to also fallback to local (like when you're on the plane and don't have network).

Essentially there is no difference between "default" and "local", other than I use "default" so that in the future we may improve on the UX if we want to. Either way, right now, the behavior of `--pull` is the same between legacy and buildkit builders.

vdemeester reviewed (approved):
Still LGTM :wink:

tiborvass commented:
Just removed `preferLocal` constant.

thaJeztah commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -114,32 +112,61 @@ func (is *imageSource) resolveLocal(refStr string) ([]byte, error) {
+		dgst, dt, err := is.resolveRemote(ctx, ref, opt.Platform)
+		// TODO: pull should fallback to local in case of failure to allow offline behavior
+		// the fallback doesn't work currently
+		return dgst, dt, err
```
nit; could be

```go
return is.resolveRemote(ctx, ref, opt.Platform)
```

tiborvass replied:
@thaJeztah true, this was because the original (desired) code is the one commented out below in which case the assignments do make sense.

thaJeztah commented on builder/builder-next/builder.go:
```diff
@@ -209,6 +209,12 @@ func (b *Builder) Build(ctx context.Context, opt backend.BuildConfig) (*builder.
 	}
 
+	if opt.Options.PullParent {
+		frontendAttrs["image-resolve-mode"] = "pull"
```
Thinking if it would be good to use the values that are / will be defined in `source.ResolveMode`

tiborvass replied:
@thaJeztah actually, could do it in a follow up once buildkit has a String() method on it. I don't want to rely on llb package at this level of abstraction.

tiborvass replied:
https://github.com/moby/buildkit/pull/577

thaJeztah replied:
Yup; definitely not a blocker (same for my other comment) 👍

thaJeztah reviewed (approved):
LGTM

triggered CI to do another run
